modules/apps/redirect/redirect-test/src/testIntegration/java/com/liferay/redirect/internal/upgrade/v3_0_2/test/UpgradeRedirectEntrySourceURLTest.java

```diff
@@ -0,0 +1,271 @@
+			"destination1", "PageA");
+		addRedirectEntry(
+			redirectEntryId2, _group1.getGroupId(), _group1.getCompanyId(),
+			"destination2", "PageB");
+		addRedirectEntry(
+			redirectEntryId3, _group1.getGroupId(), _group1.getCompanyId(),
+			"destination3", "PageC");
+
+		_redirectEntrySourceURLUpgradeProcess.upgrade();
+
+		RedirectEntry redirectEntry1 =
+			RedirectEntryLocalServiceUtil.getRedirectEntry(redirectEntryId1);
+		RedirectEntry redirectEntry2 =
+			RedirectEntryLocalServiceUtil.getRedirectEntry(redirectEntryId2);
+		RedirectEntry redirectEntry3 =
+			RedirectEntryLocalServiceUtil.getRedirectEntry(redirectEntryId3);
+
+		Assert.assertEquals("pagea", redirectEntry1.getSourceURL());
+		Assert.assertEquals("pageb", redirectEntry2.getSourceURL());
+		Assert.assertEquals("pagec", redirectEntry3.getSourceURL());
+	}
+
+	@Test
+	public void testUpgradeWithSourceURLCollision() throws Exception {
+		Assume.assumeFalse(
+			"MYSQL is case-insensitive so it is not possible to create this " +
+				"scenario, skip test.",
+			_db.getDBType() == DBType.MYSQL);
+
+		long redirectEntryId1 = _counterLocalService.increment();
+		long redirectEntryId2 = _counterLocalService.increment();
+
+		addRedirectEntry(
+			redirectEntryId1, _group1.getGroupId(), _group1.getCompanyId(),
+			"destination1", "testpage");
+		addRedirectEntry(
+			redirectEntryId2, _group1.getGroupId(), _group1.getCompanyId(),
+			"destination2", "TestPage");
+
+		_redirectEntrySourceURLUpgradeProcess.upgrade();
+
+		RedirectEntry redirectEntry1 =
+			RedirectEntryLocalServiceUtil.getRedirectEntry(redirectEntryId1);
+		RedirectEntry redirectEntry2 =
+			RedirectEntryLocalServiceUtil.getRedirectEntry(redirectEntryId2);
+
+		Assert.assertEquals("testpage", redirectEntry1.getSourceURL());
+		Assert.assertEquals("TestPage", redirectEntry2.getSourceURL());
+	}
+
+	protected void addRedirectEntry(
+			long redirectEntryId, long groupId, long companyId,
+			String destinationURL, String sourceURL)
+		throws Exception {
+
+		try (Connection connection = DataAccess.getConnection();
+			PreparedStatement preparedStatement = connection.prepareStatement(
+				StringBundler.concat(
+					"insert into RedirectEntry (uuid_, redirectEntryId,",
+					"groupId, companyId, userId, userName, createDate, ",
+					"modifiedDate, destinationURL, sourceURL) values (?, ?, ",
+					"?, ?, ?, ?, ?, ?, ?, ?)"))) {
+
+			preparedStatement.setString(1, PortalUUIDUtil.generate());
+			preparedStatement.setLong(2, redirectEntryId);
+			preparedStatement.setLong(3, groupId);
+			preparedStatement.setLong(4, companyId);
+			preparedStatement.setLong(5, TestPropsValues.getUserId());
+			preparedStatement.setString(6, null);
+			preparedStatement.setTimestamp(7, _timestamp);
+			preparedStatement.setTimestamp(8, _timestamp);
+			preparedStatement.setString(9, destinationURL);
+			preparedStatement.setString(10, sourceURL);
+
+			preparedStatement.executeUpdate();
+		}
+	}
```

Comment:
Why is this method protected? Can you make it private? Sorry I didn't see this earlier :-(

Reply:
I changed to code according to your request.